### PR TITLE
Make port arg optional in `docker compose port`

### LINKED ISSUE #11859: Make PRIVATE_PORT optional for `docker compose port`
### Description

"Normal" docker already does this. See the usage for `docker port`:

```
$ docker port --help

Usage: docker port CONTAINER [PRIVATE_PORT[/PROTO]]
```

However `docker compose` differs:
```
$ docker compose port --help

Usage:  docker compose port [OPTIONS] SERVICE PRIVATE_PORT
```

I can look into making a PR if you agree this should be optional.

### PR DESCRIPTION
**Pull request by crazybolillo**
**What I did**

This comand no longer needs a mandatory port. If no port is provided it will instead list all published ports. This behavior falls in line with the behavior provided by `docker port`.

**Related issue**
Closes #11859.

### CODE REVIEW

File: cmd/compose/port.go [hunk 1]
```diff
@@ -40,10 +41,15 @@ func portCommand(p *ProjectOptions, dockerCli command.Cli, backend api.Service)
 		ProjectOptions: p,
 	}
 	cmd := &cobra.Command{
-		Use:   "port [OPTIONS] SERVICE PRIVATE_PORT",
-		Short: "Print the public port for a port binding",
-		Args:  cobra.MinimumNArgs(2),
+		Use:   "port [OPTIONS] SERVICE [PRIVATE_PORT]",
+		Short: "List port mappings or a specific mapping for the service",
```
Comment (glours) [inline cmd/compose/port.go:45]:
```suggestion
		Short: "List port mappings or print the public port for a specific mapping for the service",
```

File: cmd/compose/port.go [hunk 2]
```diff
@@ -62,19 +68,45 @@ func portCommand(p *ProjectOptions, dockerCli command.Cli, backend api.Service)
...
-	}
 
-	fmt.Fprintf(dockerCli.Out(), "%s:%d\n", ip, port)
-	return nil
+	if opts.list {
+		summaries, err := backend.Ps(ctx, projectName, api.PsOptions{Services: []string{service}})
+		if err != nil {
+			return err
+		}
+		printPorts(dockerCli, summaries[opts.index].Publishers, opts)
```
Comment (glours) [inline cmd/compose/port.go:98]: We need to test the size of the `summaries` slice otherwise we would potentially trigger `slice bounds out of range` error

File: cmd/compose/port.go [hunk 3]
```diff
@@ -62,19 +68,45 @@ func portCommand(p *ProjectOptions, dockerCli command.Cli, backend api.Service)
...
+		return nil
+	} else {
```
Comment (glours) [inline cmd/compose/port.go:100]: the `else` isn't needed as the following statement is the last one and the `if` section finish with a `return`

**Review by glours — CHANGES_REQUESTED** [review]
Thanks for your contribution, I have some small suggestions to improve your changes

**Comment by crazybolillo** [comment]
Thanks for the review, I have taken care of your observations.

**Review by glours — APPROVED** [review]
LGTM

File: cmd/compose/port.go (hunk 1 above)
Comment (ndeloof) [reply cmd/compose/port.go]: this is just the same help used in docker/cli

File: cmd/compose/port.go [hunk 4]
```diff
@@ -40,10 +41,15 @@ func portCommand(p *ProjectOptions, dockerCli command.Cli, backend api.Service)
...
 	cmd := &cobra.Command{
-		Use:   "port [OPTIONS] SERVICE PRIVATE_PORT",
-		Short: "Print the public port for a port binding",
-		Args:  cobra.MinimumNArgs(2),
+		Use:   "port [OPTIONS] SERVICE [PRIVATE_PORT]",
+		Short: "List port mappings or print the public port of a specific mapping for the service",
+		Args:  cobra.RangeArgs(1, 2),
 		PreRunE: Adapt(func(ctx context.Context, args []string) error {
+			if len(args) == 1 {
+				opts.list = true
```
Comment (ndeloof) [inline cmd/compose/port.go:49]: incompatibility with `--index` should be tested

File: cmd/compose/port.go [hunk 5]
```diff
@@ -62,11 +68,40 @@ func portCommand(p *ProjectOptions, dockerCli command.Cli, backend api.Service)
...
+}
+
 func runPort(ctx context.Context, dockerCli command.Cli, backend api.Service, opts portOptions, service string) error {
 	projectName, err := opts.toProjectName(ctx, dockerCli)
 	if err != nil {
 		return err
 	}
+
+	if opts.list {
+		summaries, err := backend.Ps(ctx, projectName, api.PsOptions{Services: []string{service}})
```
Comment (ndeloof) [inline cmd/compose/port.go:94]: better update `backend.Port` so it can manage multiple services

File: cmd/compose/port.go [hunk 6]
```diff
@@ -62,11 +68,40 @@ func portCommand(p *ProjectOptions, dockerCli command.Cli, backend api.Service)
...
+		if err != nil {
+			return err
+		}
+		if len(summaries) <= opts.index {
```
Comment (ndeloof) [inline cmd/compose/port.go:98]: []ContainerSummary index is unrelated to replica index in service set by `--index`.

**Review by ndeloof — CHANGES_REQUESTED** [review]

File: cmd/compose/port.go [hunk 7]
```diff
@@ -62,11 +68,40 @@ func portCommand(p *ProjectOptions, dockerCli command.Cli, backend api.Service)
...
 }
 
+func printPorts(dockerCli command.Cli, publishers api.PortPublishers, opts portOptions) {
+	for _, publisher := range publishers {
+		if publisher.Protocol != opts.protocol {
+			continue
+		}
+
+		var format string
+		if strings.Contains(publisher.URL, ":") {
```
Comment (ndeloof) [inline cmd/compose/port.go:78]: would make sense to me `PortPublisher` has a `String()` method to handle this, aligned with docker/cli

**Comment by ndeloof** [comment]
while not strictly required, I think we should use this opportunity while revisiting this command to add support for app_protocol and port name (https://github.com/compose-spec/compose-spec/blob/master/05-services.md#long-syntax-3)

**Comment by crazybolillo** [comment]
> while not strictly required, I think we should use this opportunity while revisiting this command to add support for app_protocol and port name (https://github.com/compose-spec/compose-spec/blob/master/05-services.md#long-syntax-3)

Wow, did not even know those existed. How do you envision it? Allow filtering? Or just displaying the name and app_protocol if they exist for the given port?

File: cmd/compose/port.go (hunk 4 above)
Comment (crazybolillo) [reply cmd/compose/port.go:49]: Should `--index` be prohibited if no specific port was provided?

**Comment by crazybolillo** [comment]
Went ahead and changed `backend.Port `so it returns a slice of publishers. That should fix other issues like dealing with `--index`. Since I was modifying `PortPublisher` I went ahead and also changed its port members to be uint16 so it falls in line with `types.Port`

I think the only thing left is the question about `--index` incompatibility, and support for `app_protocol` and port name,

File: pkg/compose/port.go [hunk 8]
```diff
@@ -26,18 +26,40 @@ import (
 	moby "github.com/docker/docker/api/types"
 )
 
-func (s *composeService) Port(ctx context.Context, projectName string, service string, port uint16, options api.PortOptions) (string, int, error) {
+func (s *composeService) Port(ctx context.Context, projectName string, service string, port uint16, options api.PortOptions) (api.PortPublishers, error) {
```
Comment (ndeloof) [inline pkg/compose/port.go:29]: as signature changes, and breaks compatibility, makes sense to _also_ rename `Ports`(plural)

**Review by ndeloof — APPROVED** [review]

**Comment by ndeloof** [comment]
`--index` is used to select service replica, already implemented in `backend.Port` calling `getSpecifiedContainer`
Adding support for app_protocol would not be that simple, as such metadata only exists in compose model, not in actual resources. Let's keep this for a future improvement

File: pkg/compose/port.go (hunk 8 above)
Comment (crazybolillo) [reply pkg/compose/port.go:29]: @ndeloof Sorry to bother but do you have any thoughts on my comments? Should we still perform the renaming?
